From 9db497f6985ea0cc595c82deccfa2abf68233b2d Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Thu, 17 Apr 2025 20:17:27 -0700
Subject: [PATCH 05/17] win32u: Don't activate hwndTop from other non-topmost
 threads.

A hack that I don't understand really, but it allows osu! to properly lose focus on i3 again.
---
 dlls/win32u/input.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/win32u/input.c b/dlls/win32u/input.c
index 2b447f03a08..489f409a18b 100644
--- a/dlls/win32u/input.c
+++ b/dlls/win32u/input.c
@@ -2296,7 +2296,7 @@ HWND WINAPI NtUserSetFocus( HWND hwnd )
         if (call_hooks( WH_CBT, HCBT_SETFOCUS, (WPARAM)hwnd, (LPARAM)previous, 0 )) return 0;
 
         /* activate hwndTop if needed. */
-        if (!(active = get_active_window()) && !set_foreground_window( hwndTop, FALSE, FALSE )) return 0;
+        if (!(active = get_active_window()) && (get_window_long(active, GWL_EXSTYLE) & WS_EX_TOPMOST) && !set_foreground_window( hwndTop, FALSE, FALSE )) return 0;
         if (hwndTop != active)
         {
             if (!set_active_window( hwndTop, NULL, FALSE, FALSE, 0 )) return 0;
-- 
2.55.0

